core/src/test/java/io/grpc/internal/SpiffeUtilTest.java

```diff
@@ -193,4 +207,119 @@ public void spiffePathFormatTest() {
+      assertEquals("foo.bar.com", spiffeId.get().getTrustDomain());
+      assertEquals("/client/workload/1", spiffeId.get().getPath());
+    }
+
+    @Test
+    public void extractSpiffeIdFailureTest() throws CertificateParsingException {
+      Optional<SpiffeUtil.SpiffeId> spiffeId = SpiffeUtil.extractSpiffeId(serverCert0);
+      assertFalse(spiffeId.isPresent());
+      IllegalArgumentException iae = assertThrows(IllegalArgumentException.class, () -> SpiffeUtil
+          .extractSpiffeId(multipleUriSanCert));
+      assertEquals("Multiple URI SAN values found in the leaf cert.", iae.getMessage());
+
+    }
+
+    @Test
+    public void extractSpiffeIdFromChainTest() throws CertificateParsingException {
+      X509Certificate[] leafWithSpiffeChain = new X509Certificate[]{spiffeCert[0], serverCert0[0]};
+      assertTrue(SpiffeUtil.extractSpiffeId(leafWithSpiffeChain).isPresent());
+      X509Certificate[] leafWithoutSpiffeChain =
+          new X509Certificate[]{serverCert0[0], spiffeCert[0]};
+      assertFalse(SpiffeUtil.extractSpiffeId(leafWithoutSpiffeChain).isPresent());
+    }
+
+    @Test
+    public void extractSpiffeIdParameterValidityTest() {
+      NullPointerException npe = assertThrows(NullPointerException.class, () -> SpiffeUtil
+          .extractSpiffeId(null));
+      assertEquals("certChain", npe.getMessage());
+      IllegalArgumentException iae = assertThrows(IllegalArgumentException.class, () -> SpiffeUtil
+          .extractSpiffeId(new X509Certificate[]{}));
+      assertEquals("CertChain can't be empty", iae.getMessage());
+    }
+
+    @Test
+    public void loadTrustBundleFromFileSuccessTest() throws IOException,
+        CertificateParsingException {
+      SpiffeBundle tb = SpiffeUtil.loadTrustBundleFromFile(getClass().getClassLoader()
+          .getResource(TEST_DIRECTORY_PREFIX + SPIFFE_TRUST_BUNDLE_FILE).getPath());
+      assertEquals(4, tb.getSequenceNumbers().size());
+      assertEquals(123L, (long) tb.getSequenceNumbers().get("google.com"));
+      assertEquals(123L, (long) tb.getSequenceNumbers().get("test.google.com"));
+      assertEquals(12035488L, (long) tb.getSequenceNumbers().get("example.com"));
+      assertEquals(-1L, (long) tb.getSequenceNumbers().get("test.example.com"));
+      assertEquals(5, tb.getBundleMap().size());
+      assertEquals(0, tb.getBundleMap().get("google.com").size());
+      assertEquals(0, tb.getBundleMap().get("test.google.com").size());
+      assertEquals(0, tb.getBundleMap().get("test.google.com.au").size());
+      assertEquals(1, tb.getBundleMap().get("example.com").size());
+      assertEquals(2, tb.getBundleMap().get("test.example.com").size());
+      assertEquals("foo.bar.com", SpiffeUtil.extractSpiffeId(tb.getBundleMap().get("example.com")
+          .toArray(new X509Certificate[0])).get().getTrustDomain());
+    }
+
+    @Test
+    public void loadTrustBundleFromFileFailureTest() throws IOException, CertificateException {
```

Comment:
Can we either separate the different failure cases into different tests, or at the very least add comments separating them and giving a very brief description of the test case?

Reply:
Failure cases are already grouped under 2 methods - failures with reading the file itself, and failures with JSON validation. I added comments but IMHO they're repeating the related exception messages